routing.md

```diff
@@ -1,4 +1,275 @@
+
+[![state machines](collateral/routing-state-machines.png)](https://docs.google.com/spreadsheets/d/1zxm3cPZ1UDQPDpYJjGmg_qY8451WMk105HBSARJkvDI/edit#gid=0)
+
+#### Messages
+
+The only message in this protocol is the `forward` message. A simple and common version of a `forward` message might look like this:
+
+```jsonc
+{
+    "@type": "https://didcomm.org/routing/2.0/forward",
+    "to"   : "did:foo:1234abcd",
+    "payloads~attach": [
+        // One payload
+    }
+}
+```
+
+A fancier version with many optional attributes has the following potential structure:
+
+```jsonc
+{
+    "@id": "uuid value" // optional; only useful for tracing
+    "@type": "https://didcomm.org/routing/2.0/forward",
+    "to"   : "did:foo:1234abcd",
+    "payloads~attach": [
+        // An array of attached payloads, generally assumed
+        // to be DIDComm encrypted envelopes, but theoretically
+        // could be other message types as well.
+    ],
+    // This decorator and everything in it are optional. 
+    "~timing": {
+        // Usually, delay_milli and wait_until_time don't make
+        // sense together; use one or the other but not both. If
+        // both do occur, forward shouldn't happen until the later
+        // of the two conditions is satisfied.
+        "delay_milli": 12345,
+        "wait_until_time": "2020-03-25T00:00:00Z",
+        // Abandon attempt to forward after this timestamp.
+        "expires_time": "2020-03-27T18:25:00Z" 
+    },
+    // Optional: requests use of mix network instead of direct
+    // forward, to enhance privacy.
+    "mix": {
+        // Likelihood (from 0 to 1) that a random node in a
+        // mix network should be the next hop instead of
+        // sending the message directly to its final receiver.
+        "hop_chance": 0.8,
+        // If next receiver is a mix node, multiply hop_chance
+        // by this value so hop_chance attenuates for the next
+        // receiver. This prevents infinite mixing. Max value =
+        // 0.99.
+        "hop_decay": 0.2,
+        // Likelihood (from 0 to 1) that the message size will
+        // be distorted to the next receiver. 
+        "noise_chance": 0.5
+    }
+}
+```
+
+The value of the `to` field is a DID, not a key. This hides details about the internals of a sovereign domain from external parties. The sender will have had to multiplex encrypt for all relevant recipient keys, but doesn't need to know how routing happens to those keys. The mediator and the receiver may have coordinated about how distribution to individual keys takes place (see [RFC 0211: Route Coordination](https://github.com/hyperledger/aries-rfcs/blob/master/features/0211-route-coordination/README.md)), but that is outside the scope of concerns of this protocol. 
```

Comment:
using a DID as a value for the `to` field does not identify which of the agents to send the message to. How is that specified?

Reply:
Feels like maybe it should be a DID URI to an agent...`did:example:123;service=agent` ?

Reply:
I think you are conflating network routing and cryptographic routing. When Alice sends a message to Bob, she generally doesn't think of this as sending a message to Bob's cell phone specifically; she thinks of it as sending a message to Bob. I suggest that the proper cryptographic routing is to multiplex encrypt for all of Bob's agents that have the privilege of seeing plaintext (as clarified in Bob's DID doc). The message arrives at Bob's mediator attached to a `forward` request where the destination is not a specific Bob agent, but rather Bob's DID. Bob's mediator then forwards to all of Bob's agents that Bob--not Alice--has chosen to receive plaintext messages.

Reply:
Note, that authcrypt is essentially the same thing as ECDH-1PU which has pointed out that there's cocerns around key compromisation attacks. For more details on this see section 5 of the IETF spec for [ECDH-1PU](https://tools.ietf.org/html/draft-madden-jose-ecdh-1pu-00#section-5).